### PR TITLE
[API][PAP] Add close-open building endpoints

### PR DESCRIPTION
**Pull request by illusionOfParadise**

### CODE REVIEW

File: src/Pap/ClosureHistoryManager.php [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+    public function openClose(
+        string $status,
+        CampaignStatisticsOwnerInterface $objet,
+        Campaign $campaign,
+        \DateTimeInterface $doneAt,
+        Adherent $doneBy
+    ): void {
+        $stats = $objet->getStatisticForCampaign($campaign);
+        if (!$stats) {
```
Comment (elvino29) [inline src/Pap/ClosureHistoryManager.php:47]: I don't understand why you use a switch case here with only one case and no default.
```suggestion
            if (!$stats && is_a($object, Building::class) ) {
```

File: src/Controller/Api/Pap/BuildingEventController.php [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        if ($errors->count()) {
+            return $this->json($errors, Response::HTTP_BAD_REQUEST);
+        }
+
+        $entityManager->persist($buildingEvent);
+        $entityManager->flush();
+
+        $bus->dispatch(new BuildingEventCommand($building->getUuid(), $buildingEvent->getCampaign()->getUuid()));
+
+        return $this->json('OK');
```
Comment (ottaviano) [inline src/Controller/Api/Pap/BuildingEventController.php:59]:
```suggestion
        return $this->json('OK', Response::HTTP_CREATED);
```

File: src/Entity/Pap/BuildingEvent.php [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+     *     callback={"App\Pap\BuildingEventTypeEnum", "toArray"},
+     *     strict=true
+     * )
+     *
+     * @Groups({"pap_building_event_write"})
+     */
+    private ?string $type = null;
+
+    /**
+     * @ORM\Column(length=50)
```
Comment (ottaviano) [inline src/Entity/Pap/BuildingEvent.php:53]: should it be nullable? if not, what about adding `Assert\NotBlank` constraint?

File: src/Pap/BuildingEventActionEnum.php [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+namespace App\Pap;
+
+use MyCLabs\Enum\Enum;
+
+class BuildingEventActionEnum extends Enum
+{
+    public const OPEN = 'open';
+    public const CLOSE = 'close';
+
+    public const URI_PATTERN = self::OPEN.'|'.self::CLOSE;
```
Comment (ottaviano) [inline src/Pap/BuildingEventActionEnum.php:12]:
```suggestion
```

File: src/Pap/BuildingEventTypeEnum.php [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+use MyCLabs\Enum\Enum;
+
+class BuildingEventTypeEnum extends Enum
+{
+    public const BUILDING = 'building';
+    public const BUILDING_BLOCK = 'building_block';
+    public const FLOOR = 'floor';
+
+    public const URI_PATTERN = self::BUILDING.'|'.self::BUILDING_BLOCK.'|'.self::FLOOR;
```
Comment (ottaviano) [inline src/Pap/BuildingEventTypeEnum.php:13]:
```suggestion
```

File: src/Pap/Handler/BuildingEventCommandHandler.php [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    {
+        $status = BuildingEventActionEnum::CLOSE === $buildingEvent->getAction() ? BuildingStatusEnum::COMPLETED : BuildingStatusEnum::ONGOING;
+        $campaign = $buildingEvent->getCampaign();
+        /** @var CampaignStatisticsInterface $stats */
+        if (!$stats = $object->findStatisticsForCampaign($campaign)) {
+            switch (\get_class($object)) {
+                case Building::class:
+                    $stats = new BuildingStatistics($object, $campaign, $status);
+
+                    return;
```
Comment (ottaviano) [inline src/Pap/Handler/BuildingEventCommandHandler.php:87]:
```suggestion
                    break;
```

File: src/Pap/Handler/BuildingEventCommandHandler.php [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+                case Floor::class:
+                    $stats = new FloorStatistics($object, $campaign, $status);
+
+                    return;
+            }
+
+            $this->entityManager->persist($stats);
+        }
+
+        $stats->setStatus($status);
```
Comment (ottaviano) [inline src/Pap/Handler/BuildingEventCommandHandler.php:101]: useless as you already pass the status on construct

File: src/Pap/Handler/BuildingEventCommandHandler.php [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+        if (BuildingStatusEnum::COMPLETED === $status) {
+            $stats->setClosedAt($buildingEvent->getCreatedAt());
+            $stats->setClosedBy($buildingEvent->getAuthor());
+        } elseif (BuildingStatusEnum::ONGOING === $status) {
+            $stats->setClosedAt(null);
+            $stats->setClosedBy(null);
+        }
```
Comment (ottaviano) [inline src/Pap/Handler/BuildingEventCommandHandler.php:108]:
```suggestion
        $stats->setClosedAt(BuildingStatusEnum::COMPLETED === $status ? $buildingEvent->getCreatedAt() : null);
        $stats->setClosedBy(BuildingStatusEnum::COMPLETED === $status ? $buildingEvent->getAuthor() : null);
```

**Review by ottaviano — APPROVED** [review]

**Review by elvino29 — APPROVED** [review]

**Pull request merged**